nifi-nar-bundles/nifi-smb-bundle/nifi-smb-processors/src/main/java/org/apache/nifi/processors/smb/FetchSmb.java

```diff
@@ -0,0 +1,230 @@
+                    .description("A flowfile will be routed here for each successfully fetched File.")
+                    .build();
+
+    public static final Relationship REL_FAILURE =
+            new Relationship.Builder().name("failure")
+                    .description(
+                            "A flowfile will be routed here for each File for which fetch was attempted but failed.")
+                    .build();
+
+    public static final Relationship REL_INPUT_FAILURE =
+            new Relationship.Builder().name("input_failure")
+                    .description("The incoming flowfile will be routed here if its content could not be processed.")
+                    .build();
+
+    public static final Set<Relationship> relationships = Collections.unmodifiableSet(new HashSet<>(asList(
+            REL_SUCCESS,
+            REL_FAILURE,
+            REL_INPUT_FAILURE
+    )));
+
+    private static final List<PropertyDescriptor> PROPERTIES = asList(
+            REMOTE_FILE,
+            REMOTE_FILE_FIELD,
+            SMB_CLIENT_PROVIDER_SERVICE,
+            RECORD_READER
+    );
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        boolean shouldDeleteFlowFile = false;
+
+        final SmbClientProviderService clientProviderService =
+                context.getProperty(SMB_CLIENT_PROVIDER_SERVICE).asControllerService(SmbClientProviderService.class);
+
+        try (SmbClientService client = clientProviderService.getClient()) {
+
+            if (context.getProperty(RECORD_READER).isSet()) {
+                final RecordReaderFactory recordReaderFactory =
+                        context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+
+                try (InputStream inFlowFile = session.read(flowFile);
+                        RecordReader reader = recordReaderFactory.createRecordReader(flowFile, inFlowFile, getLogger())
+                ) {
+                    final String fieldName = context.getProperty(REMOTE_FILE_FIELD).getValue();
+                    Record record;
+                    while ((record = reader.nextRecord()) != null) {
+                        final String fileName = record.getAsString(fieldName);
+                        final FlowFile outFlowFile = session.create(flowFile);
+                        fetchAndTransfer(session, client, outFlowFile, fileName);
+                    }
+                    shouldDeleteFlowFile = true;
+                } catch (Exception e) {
+                    handleInputError(e, session, flowFile);
+                }
```

Comment:
In case of "input error", the session needs to be rolled back I think, because some FFs may already be processed and sent to SUCCESS/FAILURE. After rollback, the original FF can be sent to REL_INPUT_FAILURE. In "input error" case, I think only this output needs to be produced.